fix tests, optimize contract

skromez opened the pull request:
```solidity
    // PhABC: Don't think this function is very useful, I would get rid of it.
    function prizeSent(address account, uint256 _giveawayId) public {
        Giveaway storage _giveaway = giveawaysMap[account][_giveawayId];
        require(_giveaway.finished == true, "giveaway is not finished");
        require(msg.sender == account, "only host of giveaway can call prizeSent");
        require(_giveaway.winner != address(0), "winner should not be address zero");
        _giveaway.prizeSent = true;
    }
```

^ about this, I still think it's might be needed, mostly for UI/UX purposes, cuz after the giveaway is over, owner of giveaway need to send the items out. On UI there's button "Send Prizes" and inside of it it calls this as well. And after that on UX I checking this prizeSent variable to either hide/show this button and other things for winner screen.

skromez committed:
fix tests, optimize contract. thanks PhABC!

skromez committed:
change variable name

PhABC commented on hardhat/contracts/AirChrysalis.sol:
```diff
@@ -54,74 +51,62 @@ contract AirChrysalis is VRFConsumerBaseV2 {
-        giveawayId++;
+        giveaways.push(Giveaway(_contractAddr, _tokensTuple, participants, false, address(0), msg.sender, 0));
+        addressGiveaways[msg.sender].push(giveaways.length);
+        emit GiveawayCreated(msg.sender, giveaways.length);
```
Do we need a global place to track giveaways? 

I would remove `giveaways` and just do 

```ts
addressGiveaways[msg.sender].push(Giveaway(_contractAddr, _tokensTuple, participants, false, address(0), msg.sender, 0));
emit GiveawayCreated(msg.sender, addressGiveaways[msg.sender].length);
```

This would save a bit more than 20k gas

skromez committed:
add multiple winners implementation, fix tests

skromez commented:
Quick update on what I've changed in the contract to support multiple winners.
1) added uint32 winnersAmount variable 
2) added address owner variable for require(...) check since we got rid of some global variables where we stored that.
3) added uint256 requestId and uint256[] randomWords which I'd use to calculate and determine who are the winners of the giveaway. I thought about it for a little bit and came up with an approach where we don't even need calculate/store winners in the contract for each giveaway, we only need randomNumbers.

How are we gonna determine who are the winners of certain giveaway?
For example:

We have 10 participants and winnersAmount = 3, and we got 3 randomNumbers from Chainlink VRF, let them be randomNumbers = [315123, 81238, 512361]. We could actually calculate who are the winners are on front-end with given input and we don't need to store it anywhere in contract. 

```js
const participants = [1,2,3,4,5,6,7,8,9,10];
const winnersAmount = 3;
const randomNumbers = [315123, 81238, 512361];
const winners = [];

for (let i = 0; i < winnersAmount; i++) {
    const rndNumber = randomNumbers[i];
    const winnerIdx = rndNumber % participants.length;
    winners.push(participants[winnerIdx]);
    participants.splice(winnerIdx, 1);
}
console.log(winners) // [4, 6, 2] - this result will be the same as long as randomNumbers are [315123, 81238, 512361]
```
this code would always return the same winners if on randomNumbers are the same.

Also for multiple winners, I thought it would make sense to be able to create giveaway only if prizes could be evenly split between winnersAmount. That's why I included this check 
```solidity
if (winnersAmount > 1) {
   require(_tokensTuple.length % winnersAmount == 0, "amount of items should be divided to winners amount without remainder");
}
```

And after we determine who are the winners of giveaway on front-end we can split the prizes this way
```js
const items = [1,2,3,4,5,6,7,8,9,10];
const winners = [1,2];
const itemsAmountToEachWinner = items.length / winners.length;
const winnerItems = []
let offset = 0;

while (offset != items.length) {
    winnerItems.push(items.slice(offset, offset + itemsAmountToEachWinner))
    offset += itemsAmountToEachWinner
}

for (let i = 0; i < winners.length; i++) {
    const winner = winners[i]
    const itemsForWinner = winnerItems[i]
    // 1st iteration winner = 1 and prizes for him will be first 5 items from prizes.
    // 2nd iteration winner = 2 and prizes for him will be second 5 items from prizes.
}
```
This is not the final solution on front-end, but wanted to get your opinion on this approach.

PhABC commented:
Looking very good so far! One small optimization you could make is instead of requesting N random values from Chainlink VRF, you can just request one and then derive more random number for it by doing onion hashing, e.g. 

```js
...
bytes32 randseed = keccack256(vrf(...));
uint256[] randomNumbers = [];

for (uint i = 0; i < n; i++) {
  randomNumbers.push(uint256(randSeed) % n);
  randseed = keccak256(randseed);
 }
 ...
 ```
This is deterministic and will allow you to query and store a single RNG number, which will be much better than say 1000 if there was 1000 winners.

skromez commented:
> Looking very good so far! One small optimization you could make is instead of requesting N random values from Chainlink VRF, you can just request one and then derive more random number for it by doing onion hashing, e.g.
> 
> ```js
> ...
> bytes32 randseed = keccack256(vrf(...));
> uint256[] randomNumbers = [];
> 
> for (uint i = 0; i < n; i++) {
>   randomNumbers.push(uint256(randSeed) % n);
>   randseed = keccak256(randseed);
>  }
>  ...
> ```
> 
> This is deterministic and will allow you to query and store a single RNG number, which will be much better than say 1000 if there was 1000 winners.

Thank you for showing this snippet, definitely would save a lot of gas on requesting less things from VRF.
Will implement this!

skromez committed:
contract changes, front-end changes
